### PR TITLE
Create CODE_OF_CONDUCT.md

### PR DESCRIPTION
**Pull request by philberty**
Adding a code of conduct lays out rules that we can all follow in the project.

Although GCC does not have one of these we can at least adopt one here for the project development.

### CODE REVIEW

**Review by flip1995 — APPROVED** [review]
With CoCs involved it is important to note, that those are mostly just common sense. The TL;DR is "Don't be an a**". This does not restrict anyone from contributing or expressing their opinion, but guards Rust-GCC from malicious actors from outside the project, since we can point them to the CoC without them using an "I didn't know" excuse.

I'm definitely in favor of this!

**Comment by bjorn3** [comment]
I am in favor of having a code of conduct. This PR adapts the contributor convenant as code of conduct, but there are many other code of conduct like documents like the [rust](https://www.rust-lang.org/policies/code-of-conduct) code of conduct or the [gnu](https://www.gnu.org/philosophy/kind-communication.html) kind communication guidelines. While I do not have any objections to the contributor convenant, I think it is worthwhile to investigate alternative code of conducts. Any problems with the chosen code of conduct could be used by community leaders to wrongly reject people (has happened before. eg https://www.fast.ai/2020/10/28/code-of-conduct/) or by contributors to justify inappropriate behavior.

**Comment by dkm** [comment]
there's also the work done by some GNU hackers here:
https://wiki.gnu.tools/gnu:social-contract

A new website (https://gnu.tools/ ) will be pushed online today I think, so we may want to read this before deciding, maybe GCCrs could use it.

**Comment by philberty** [comment]
Sounds good @dkm there is no rush to merge this so we can wait.

**Comment by flip1995** [comment]
IIUC the GNU assembly is also just adopting the CCCoC. The social contract is not a CoC, but rather the Vision of the GNU assembly project.

**Comment by dkm** [comment]
> IIUC the GNU assembly is also just adopting the CCCoC. The social contract is not a CoC, but rather the Vision of the GNU assembly project.

Yes you are right, thanks for correcting me !

**Comment by NalaGinrut** [comment]
Looks great to me!
For any inappropriate words, I think it's better to publicly correct first. Try to not to blame first. This will be helpful to make the community more friendly and attractive.

File: CODE_OF_CONDUCT.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+## Enforcement
+
+Instances of abusive, harassing, or otherwise unacceptable behavior may be
+reported to the community leaders responsible for enforcement at
+https://gcc-rust.zulipchat.com/.
+All complaints will be reviewed and investigated promptly and fairly.
+
+All community leaders are obligated to respect the privacy and security of the
+reporter of any incident.
```
Comment (tschwinge) [inline CODE_OF_CONDUCT.md:67]: As it requires registration, and messages by default are public, does the Zulip chat suffice as point of contact (`[INSERT CONTACT METHOD]` in the original CCCoC v2.0 text), in particular the _[obligation] to respect the privacy [...]_?  Should this maybe rather be an email address?  What are other software communities putting there?

**Comment by tschwinge** [comment]
A few "random" comments.  ;-)

I personally always wish that one doesn't need (to actually make use of) these formal rules, and instead everybody just behave reasonable by default -- but, oh well, reality teaches us...  (No issues in this group here, though, as far as I can tell.)  :-)

So, I'm not objecting to adapting this, or another set of reasonable rules.

GCC (and GNU generally) don't publish a CoC, instead GNU proposes "GNU Kind Communications Guidelines", <https://www.gnu.org/philosophy/kind-communication.html>, <https://lwn.net/Articles/769167/>.

Just as one other data point: the GNU Tools Cauldron 2019, as an in-person meeting, did have a CoC: <https://gcc.gnu.org/wiki/cauldron2019#Code_of_Conduct>, _British Computer Society Open Source Special-ist Group Code of Conduct, <https://ossg.bcs.org/coc/>_.

**Comment by philberty** [comment]
Thanks, @tschwinge, I found this also https://gnu.tools/en/documents/code-of-conduct/ which might be more appropriate. Github encourages projects to have a contributor guideline and CoC in the security settings iirc.

I raised this PR to make a discussion more than anything if we think we don't need this then don't have to merge it.

**Comment by flip1995** [comment]
To summarize the discussion: 

- We want a CoC.
- There are multiple different CoCs out there 
- Each CoC might have its pros and cons
- Suggested CoCs:
    - [Contributor Covenant Code of Conduct](https://www.contributor-covenant.org/) (used by GNU Assembly)
    - [Rust Code of Conduct](https://www.rust-lang.org/policies/code-of-conduct)
    - [GNU Kind Communications Guidelines](https://www.gnu.org/philosophy/kind-communication.html)
    - [BCS Code of Conduct](https://ossg.bcs.org/coc/)

Should we vote on which CoC we should adopt? How should we go ahead and decide this? Next community meeting?

---

<details>
<summary>Now facts end and my opinion starts:</summary>

The CCCoC is a pretty widely adopted CoC and I don't really see a problem why we couldn't adopt it. The problems that arose from this CoC, that  @bjorn3 pointed out (https://www.fast.ai/2020/10/28/code-of-conduct/) came mostly from the people applying the CoC rather than from the CoC itself (IMO). 

The Rust CoC would also make sense in a modified form, since this is also kind of a Rust project. Personally I like the Rust CoC the most, since it does the best job at communicating what is and isn't allowed in the community.

The GNU Kind Communications Guidelines would make sense, since this should be upstreamed to GCC in the future.

The BCS CoC would make the least sense IMO, since the BCS isn't related to this project.
</details>

**Comment by tschwinge** [comment]
Additional item to resolve: point of contact, <https://github.com/Rust-GCC/gccrs/pull/372#pullrequestreview-639212843>.

---

I agree that the BCS CoC makes the least sense; I just mentioned that one for completeness, as is appeared in the GNU Tools (Cauldron) context.

I have no strong preference between the other ones.

---

> The GNU Kind Communications Guidelines would make sense, since this should be upstreamed to GCC in the future.

In some way, yes, but on the other hand: the CoC is not about the source code itself (that's what is to be upstreamed) but it's rather about the community of people (which ideally at some point also should "move upstream"), so in my opinion, there's no problem if GCC upstream and GCC/Rust are governed by different CoCs.

---

> Should we vote on which CoC we should adopt? How should we go ahead and decide this? Next community meeting?

Yeah, why not.

Thus, anyone who's not able to attend the next meeting, should be sure to voice their opinions before that one.

**Comment by bjorn3** [comment]
> * [Contributor Covenant Code of Conduct](https://www.contributor-covenant.org/) (used by GNU Assembly)
> 
> * [Rust Code of Conduct](https://www.rust-lang.org/policies/code-of-conduct)
> 
> * [GNU Kind Communications Guidelines](https://www.gnu.org/philosophy/kind-communication.html)

I am fine with all of them.

**Comment by philberty** [comment]
As per our last online call, I have updated the contact details to point to the organisation members, and I will be merging this. We can always change if this doesn't suit our project so please feel free to discuss this further.

**Comment by tschwinge** [comment]
> Although GCC does not have [a code of conduct]

As of last week, it now does: <https://inbox.sourceware.org/CADzB+2=V=dfUK3caGE6rjYPw93NWnvyAT8vFn6bf-KoPd33bKQ@mail.gmail.com> "Announcing GCC Code of Conduct".  :tada:

> we can at least adopt one here for the project development.

Which we then did.

Should we now switch to the upstream "GCC Code of Conduct", <https://gcc.gnu.org/conduct.html>?  (..., which does seem very reasonable to me.)  I'd offer to implement the necessary (few) changes in GCC/Rust.

Maybe a discussion item for the next monthly GCC/Rust call?

**Pull request merged**